File: 187-repeated-dna-sequences/187.cpp
#include <iostream>
#include <vector>
#include <unordered_set>

using namespace std;

class Solution {
public:
    vector<string> findRepeatedDnaSequences(string s) {
        int len = s.length();
        if (len < 11) return {};
        unordered_set<string> one, second;
        one.insert(s.substr(0, 10));
        for (int i = 1; i < len-9; ++i) {
            string tmp = s.substr(i, 10);
            if (one.count(tmp)) {
                second.insert(tmp);
            }
            else {
                one.insert(tmp);
            }
        }
        return vector<string>(second.begin(), second.end());
    }
};

int main(int argc, char** args) {
    Solution sol = Solution();
    vector<string> res1 = sol.findRepeatedDnaSequences("AAAAACCCCCAAAAACCCCCCAAAAAGGGTTT");
    for (auto a : res1) {
        cout << a << " ";
    }
    cout << endl;

    vector<string> res2 = sol.findRepeatedDnaSequences("AAAAAAAAAAAAA");
    for (auto a : res2) {
        cout << a << " ";
    }
    cout << endl;

    return 0;
}

File: 187-repeated-dna-sequences/187.py
#!/usr/bin/env python
# -*- coding:utf-8 -*-

from typing import List

class Solution:
    def findRepeatedDnaSequences(self, s: str) -> List[str]:
        if len(s) <= 10:
            return []
        one, second = set(), set()
        one.add(s[:10])
        for i in range(1, len(s)-9):
            if s[i:i+10] in one:
                second.add(s[i:i+10])
            else:
                one.add(s[i:i+10])
        return list(second)


if __name__ == '__main__':
    sol = Solution()
    print(sol.findRepeatedDnaSequences('AAAAACCCCCAAAAACCCCCCAAAAAGGGTTT'))
    print(sol.findRepeatedDnaSequences('AAAAAAAAAAAAA'))